# Linear regression with multiple features/variables
Let's extend the idea of linear regression to work with multiple independent variables (Multivariate Linear Regression).

## Problem context:
We want to sell houses and based on some characteristics you want to know what a good market price would be.
We have gathered information on the recent houses that were sold in the area.

It is your job to predict housing prices based on these features.

The file ex1data2.txt contains a training set of housing prices in Portland, Oregon.

The data is organized as
- 1st column : size of the house (in square feet),
- 2nd column is the number of bedrooms,
- 3rd column is the price of the house.

The only difference with the previous example is that we now have more than one independent variables (but the concepts you have learnt in the previous section applies here as well).

In [86]:
# #################################################
# FILL IN THE NECESSARY CODE.
# Import the necessary packages
# ps: We need to be able to manipulate dataframes and perform matrix calculation
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [87]:
# #################################################
# FILL IN THE NECESSARY CODE.
# Read the data from the file "ex1data2.txt" (available on Canvas) and display the top 5 rows of this data.
data = pd.read_csv('Data/ex1data2.txt', header = None, delimiter = ",") #read from dataset
X = data.iloc[:,0:2] # read first column, will be put in a 'series' variable
y = data.iloc[:,2] # read second column, will be put in a 'series' variable
m = len(y) # number of training example (97)
print(data.head()) # view first few rows of the data

      0  1       2
0  2104  3  399900
1  1600  3  329900
2  2400  3  369000
3  1416  2  232000
4  3000  4  539900


Fill in the following. Make sure that we have numpy arrays to continue our calculations

In [88]:
# #################################################
# FILL IN THE NECESSARY CODE.
# We need
# print(X)
# print(y)
X = X.to_numpy(dtype=float)
y1 = y.to_numpy()[:, np.newaxis]
# y2 = y.to_numpy()
# print(X.shape)
# print(y1.shape)
# print(y2.shape)
m = len(X)

## Feature Normalization
Looking at the data, the house sizes are about 1000 times the number of bedrooms. 
So the house size is orders of magnitude larger, so the best is to "scale the features" or "normalize", 
this will have as benefit that gradient descent converge much more quickly.

We will
  - calculate the mean of each of the features (Xo, X1)
  - subtract the mean value of each feature from the dataset.
  - divide each feature values by their respective standard deviations

![axis in numpy](./images/axis_0_axis_1_in_numpy.jpg)

In [89]:
# #################################################
# FILL IN THE NECESSARY CODE.
# Normalize X
# Should we normalize y?
Xmean = np.mean(X, axis=0)
Xstd = np.std(X, axis=0)
print(Xmean.shape)
print(Xmean)
print(Xstd.shape)
print(Xstd)
X[:, 0] = (X[:, 0] - Xmean[0]) / Xstd[0]
X[:, 1] = (X[:, 1] - Xmean[1]) / Xstd[1]

(2,)
[2000.68085106    3.17021277]
(2,)
[7.86202619e+02 7.52842809e-01]


Adding the intercept term and initializing parameters

In [91]:
# #################################################
# FILL IN THE NECESSARY CODE.
# Augment the matrix X so that we can perform a X*theta in 1 go. 
# So include the possibility to multiply in 1 go, including theta0.
# Convert X_norm into the variable X_aug
# So include the possibility to multiply X*theta in 1 go, including theta0 (the intercept).
theta = np.zeros([3, 1])
ones = np.ones((m, 1))
X_aug = np.column_stack((ones, X))
print(X_aug.shape)
print(len(X_aug))
print(X_aug[0:3])

(47, 3)
47
[[ 1.          0.13141542 -0.22609337]
 [ 1.         -0.5096407  -0.22609337]
 [ 1.          0.5079087  -0.22609337]]


In [140]:
# #################################################
# FILL IN THE NECESSARY CODE.
# Initialize all the hyperparameters of our model
#   Use alpha = for the learning rate
#   num_iters = number of iterations
#   theta : initialize  theta
alpha = 0.008
num_iters = 10_000
# theta already initialized to zero

Computing the cost

In [141]:
print(X_aug.shape)
print(theta.shape)

(47, 3)
(3, 1)


In [142]:
# #################################################
# FILL IN THE NECESSARY CODE.
# Calculate the MSE.
def computeCostMulti(X, y, theta):
    temp = np.dot(X, theta) - y
    return np.sum(np.power(temp, 2)) / (2 * m)

In [143]:
J = computeCostMulti(X_aug, y, theta)
print(J)

65591548106.45744


<b>Check your calculation</b>

Cost :  65591548106.45744

## Finding the optimal parameters using Gradient Descent

In [144]:
# #################################################
# FILL IN THE NECESSARY CODE.
#    X contains the multivaraite input
#    y contain the labels
#    theta : parameters list of our linear model
#    alpha : the learning rate
#    num_iters : number of iterations
#    return parameter (theta) : parameters list of our linear model

def gradientDescentMulti(X, y, theta, alpha, iterations):
    for _ in range(iterations):
        temp = np.dot(X, theta) - y
        temp = np.dot(X.T, temp)
        theta = theta - (alpha/m) * temp
    return theta

In [145]:
theta = gradientDescentMulti(X_aug, y, theta, alpha, num_iters)
# print(theta)
iter_values = [100, 200, 400, 1000, 5000, 10000]

for iteration in iter_values:
    theta0 = np.zeros([3, 1])
    theta_result = gradientDescentMulti(X_aug, y, theta0, 0.03, iteration)
    print(f"Theta with {iteration} iterations: {theta_result[0]}, {theta_result[1]}, {theta_result[2]}")

Theta with 100 iterations: [324225.18388214], [93661.31403812], [8355.5486845]
Theta with 200 iterations: [339642.90450832], [105377.32631522], [-2514.95084841]
Theta with 400 iterations: [340410.91897274], [109162.68848142], [-6293.24735132]
Theta with 1000 iterations: [340412.65957445], [109447.698242], [-6578.25662652]
Theta with 5000 iterations: [340412.65957447], [109447.79646964], [-6578.35485416]
Theta with 10000 iterations: [340412.65957447], [109447.79646964], [-6578.35485416]


<pre>
(results with alpha = 0.03)
Theta with    100 iterations : [[3.39642905e+05]  [1.44952956e+05]  [5.98800848e+01]]
Theta with    200 iterations : [[3.39642905e+05]  [1.44952956e+05]  [5.98800848e+01]]
Theta with    400 iterations : [[3.40410919e+05]  [1.53263978e+05]  [4.64918053e+01]]
Theta with   1000 iterations : [[3.40412660e+05]  [1.53769414e+05]  [-6.77111420e+00]]
Theta with   5000 iterations : [[340412.65957447] [153769.70114155] [-363.65654995]]
Theta with  10000 iterations : [[340412.65957447] [153769.94033767] [-809.74547603]]
</pre>
We now have the optimized value of theta. Use these values in the above cost function.

In [127]:
for iteration in iter_values:
    theta0 = np.zeros([3, 1])
    theta_result = gradientDescentMulti(X_aug, y, theta0, 0.03, iteration)
    J = computeCostMulti(X_aug, y, theta_result)
    print(f"Cost with {iteration} iterations: {J}")
# J = computeCostMulti(X_aug, y, theta)
# print(J)

Cost with 100 iterations: 2278400086.9068036
Cost with 200 iterations: 2050854463.9196255
Cost with 400 iterations: 2043315820.8112059
Cost with 1000 iterations: 2043280050.6070747
Cost with 5000 iterations: 2043280050.6028283
Cost with 10000 iterations: 2043280050.6028283


<pre>
Cost with    100 iterations :   10596969344.166979
Cost with    200 iterations :   3344770635.4916563
Cost with    400 iterations :   2105448288.6292474
Cost with   1000 iterations :   2043498948.1433077
Cost with   5000 iterations :   2043280050.6028287
Cost with  10000 iterations :   2043280050.6028283
</pre>

## Interference Step
What would be the price for a 2480 square foot, 3 bedroom house?

How confident are you about this prediction?

In [ ]:
# #################################################
# FILL IN THE NECESSARY CODE.

What would be the price for a 5500 square foot, 6 bedroom house?

How confident are you about this prediction?

In [ ]:
# #################################################
# FILL IN THE NECESSARY CODE.